### PR TITLE
Enable anonymised database dump and restore

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
Provide means to dump and restore database

#### Ticket

[CBO-1252](https://dsdmoj.atlassian.net/browse/CBO-1252)

#### Why
So we can debug production problems including
MI report failure

### CODE REVIEW

File: kubernetes_deploy/jobs/migrate.yaml [hunk 1]
```diff
@@ -24,13 +24,22 @@ spec:
...
+            - name: ADVOCATE_PASSWORD
+              value: just-be-present
+            - name: ADMIN_PASSWORD
+              value: just-be-present
+            - name: SUPERADMIN_USERNAME
+              value: just-be-present
+            - name: SUPERADMIN_PASSWORD
+              value: just-be-present
+            - name: TEST_CHAMBER_API_KEY
+              value: just-be-present
```
Comment (jsugarman) [inline kubernetes_deploy/jobs/migrate.yaml]: remove all but secret key base

File: kubernetes_deploy/jobs/seed.yaml [hunk 2]
```diff
@@ -24,43 +24,22 @@ spec:
...
-                  key: SECRET_KEY_BASE
+                  key: AWS_REGION
+            - name: SECRET_KEY_BASE
+              value: just-be-present
             - name: CASE_WORKER_PASSWORD
-              valueFrom:
-                secretKeyRef:
-                  name: cccd-secrets
-                  key: CASE_WORKER_PASSWORD
+              value: just-be-present
```
Comment (jsugarman) [inline kubernetes_deploy/jobs/seed.yaml]: Just use `ConfigMapRef` and `SecretRef` for simplicity

File: lib/tasks/dump.rake [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+    def sensitive_table_exclusions
+      # make sure a db:dump task exists for each of the excluded tables (i.e. db:dump:providers)
+      excluded_tables.map { |table| "--exclude-table-data #{table}" }.join(' ')
+    end
+
+    def exec_with_output(cmd)
+      success = system(cmd)
+      success ? puts('done'.green) : puts('failed'.red)
+    end
```
Comment (jsugarman) [inline lib/tasks/dump.rake]: not used - remove

File: lib/tasks/rake_helpers/rake_utils.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  def decompress_file(filename)
+    shell_working "decompressing file #{filename}" do
+      system "gunzip -3 -f #{filename}"
+    end
+  end
+
+  def production_protected
+    raise 'This operation was aborted because the result might destroy production data' if Rails.host.production?
+  end
+end
```
Comment (jsugarman) [inline lib/tasks/rake_helpers/rake_utils.rb]: add blank line

File: spec/rails_helper.rb [hunk 5]
```diff
@@ -28,6 +28,7 @@
   # exclude patterns from test coverage results
   add_filter %r{^/lib\/rack/} # only used to prevent feature test flickering
   add_filter %r{^/lib\/demo_data/} # only used for generation of demo data
+  add_filter %r{^/lib\/tasks/} # only used for generation of demo data
```
Comment (jsugarman) [inline spec/rails_helper.rb]: amend or remove comment

File: kubernetes_deploy/scripts/job.sh [hunk 6]
```diff
@@ -3,8 +3,8 @@ function _job() {
   usage="job -- run job in the specified environment
```
Comment (kmahern) [inline kubernetes_deploy/scripts/job.sh:3]: context='live-1' (line 45) fails for me (not something introduced by this change though) - should be context='live-1.cloud-platform.service.justice.gov.uk' to make it generic for all devs?

**Review by kmahern — COMMENTED** [review]
Looks great, but current (documented) solution only really works for people with access to the CCCD secrets. Do we either need to commit to setting up all devs investigating incidents in production with access to secrets (with time and risk this involves) or workaround solution?

File: kubernetes_deploy/scripts/job.sh (hunk 6 above)
Comment (jsugarman) [reply kubernetes_deploy/scripts/job.sh:3]: @kmahern I have modified to use the users `--current` context and just changes `--namespace` - can you give the job a go again and let me know if that fixes the issues

**Comment by jsugarman** [comment]
> Looks great, but current (documented) solution only really works for people with access to the CCCD secrets. Do we either need to commit to setting up all devs investigating incidents in production with access to secrets (with time and risk this involves) or workaround solution?

So, for your purposes we need to add you to git-crypt  - can you send me your key. For devs without secrets access a dev with secrets access could restore the database for them on an environment, which we should do anyway once this is in. This needs testing but i believe it should be possible to go onto the environment and run the `db:dump:copy_s3_dump['s3_key_for_file']` (but there may be a need to provide credentials for the bucket actually) and then `db:restore['downloaded-dump-file']`.

This could be a separate PR and bit of work which could also do with separate documentation

File: kubernetes_deploy/scripts/job.sh (hunk 6 above)
Comment (kmahern) [reply kubernetes_deploy/scripts/job.sh:3]: 👍 yes this worked.

**Review by kmahern — APPROVED** [review]
great 👍

**Pull request merged**